code/src/sixsq/nuvla/server/resources/api_docs/openapi.clj

```diff
@@ -0,0 +1,122 @@
+            [sixsq.nuvla.server.resources.api-docs.openapi-schema :as openapi-schema]
+            [spec-tools.openapi.core :as openapi]))
+
+
+(defn tags
+  "Returns the `tags` section of the OpenAPI spec."
+  [resources-metadata]
+  (->> resources-metadata
+       (map #(select-keys % [:name :description]))
+       (sort-by :name)))
+
+
+(defn resource-get
+  "Returns the OpenAPI method spec for the retrieve operation"
+  [{resource-type-name :name :as resource-metadata}]
+  {:get
+   {:tags        [resource-type-name]
+    :description (str "Returns a " resource-type-name)
+    :responses   {200 {:description (str "A " resource-type-name)
+                       :content
+                       {"application/json"
+                        {:schema {:$ref (str "#/components/schemas/"
+                                             (openapi-schema/resource-schema-name resource-metadata))}}
+                        ;; TODO: generate examples ?
+                        ;; :examples {}
+                        }}
+                  403 {:description "Not authorized"}
+                  404 {:description "The resource was not found"}}}})
+
+
+(defn collection-query
+  "Returns the OpenAPI method spec for the query operation"
+  [{resource-type-name :name :as _resource-metadata}]
+  {:get
+   {:tags        [resource-type-name]
+    :description (str "Returns a collection of " resource-type-name)
+    :responses   {}}})
+
+
+(defn resource-paths
+  "Returns the `paths` section of the OpenAPI spec for the specified resource."
+  [{:keys [type-uri scrud-operations] resource-type-name :name :as resource-metadata}]
+  (let [scrud-op-supported?  (set scrud-operations)
+        resource-param "resource-id"
+        resource-ops   (merge
+                         (when (scrud-op-supported? :get)
+                           (resource-get resource-metadata)))
+        collection-ops (merge
+                         (when (scrud-op-supported? :query)
+                           (collection-query resource-metadata)))]
+    (merge
+      (when (seq collection-ops)
+        {(str "/" type-uri) collection-ops})
+      (when (seq resource-ops)
+        {(str "/" type-uri "/{" resource-param "}")
+         (merge
+           {:parameters [{:name        resource-param
+                          :in          "path"
+                          :required    true
+                          :description (str "The ID of the " resource-type-name)
+                          :schema      {:type   "string"
+                                        :format "uuid"}}]}
+           resource-ops)}))))
+
+
+(defn paths
+  "Returns the `paths` section of the OpenAPI spec."
+  [resources-metadata]
+  (reduce (fn [m resource-metadata]
+            (merge m (resource-paths resource-metadata)))
+          (sorted-map)
+          resources-metadata))
+
+
+(defn schemas
+  "Returns the `schemas` section of the OpenAPI spec."
```

Comment:
One schema per resource type is generated. These schemas are then references from the `paths` section of the spec. Currently there is no attempt at reusing common schema parts,so common properties are duplicated across all resource types.